src/Voron/Impl/Journal/WriteAheadJournal.cs

```diff
@@ -953,53 +943,80 @@ public LockTaskResponsible(object @lock, CancellationToken token)
-                    var isLockTaken = false;
+                    var current = new AssignedTask(task);
                     try
                     {
-                        _waitForTaskToBeDone.Reset();
-                        _task += task;
-                        do
+                        while (true)
+                        {
+                            var isAssigned = Interlocked.CompareExchange(ref _active, current, null) == null;
+                            if (isAssigned)
+                                break;
+
+                            if (_waitForTaskToBeDone.Wait(TimeSpan.FromMilliseconds(250), _token))
+                            {
+                                _waitForTaskToBeDone.Reset();
+                            }
+                        }
+
+                        while (true)
                         {
+                            var isLockTaken = false;
                             Monitor.TryEnter(_lock, 0, ref isLockTaken);
                             if (isLockTaken)
                             {
-                                RunTaskIfNotAlreadyRan();
-                                break;
+                                try
+                                {
+                                    task();
+                                    return true;
+                                }
+                                finally
+                                {
+                                    Monitor.Exit(_lock);
+                                }
                             }
 
-                            try
+                            if (_waitForTaskToBeDone.Wait(TimeSpan.FromMilliseconds(250), _token))
                             {
-                                if (_waitForTaskToBeDone.Wait(TimeSpan.FromMilliseconds(250), _token))
-                                    break;
+                                _waitForTaskToBeDone.Reset();
                             }
-                            catch (OperationCanceledException)
+
+                            if (current.DoneFlag.IsRaised())
                             {
-                                _task = null;
-                                return false;
+                                current.Error?.Throw();
+                                return true;
                             }
-                        } while (_task != null);
-
-                        return true;
+                        }
                     }
-                    finally
+                    catch (OperationCanceledException)
                     {
-                        if (isLockTaken)
-                            Monitor.Exit(_lock);
+                        return false;
                     }
                 }
 
                 public void RunTaskIfNotAlreadyRan()
                 {
-                    if (_task == null)
-                        return;
+                    Debug.Assert(Monitor.IsEntered(_lock));
```

Comment:
For now, the implementation is every call to  `RunTaskIfNotAlreadyRan` only one assigned task executes.
Is it make sense to change it to run all the assigned tasks?

Reply:
No, just have a single one, we'll re-execute it soon anyway